// api/dist/apps/api/src/modules/bookings/bookings.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BookingsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const shared_1 = require("@waterting/shared");
const audit_service_1 = require("../../common/audit/audit.service");
let BookingsService = class BookingsService {
    constructor(prisma, audit) {
        this.prisma = prisma;
        this.audit = audit;
    }
    async create(user, data) {
        return this.prisma.$transaction(async (prisma) => {
            const unit = await prisma.unit.findFirst({
                where: { id: data.unitId },
                include: { tower: { include: { project: true } } },
            });
            if (!unit || unit.tower.project.tenantId !== user.tenantId) {
                throw new common_1.NotFoundException('Unit not found');
            }
            if (unit.status !== shared_1.UnitStatus.AVAILABLE && unit.status !== shared_1.UnitStatus.RESERVED) {
                throw new common_1.BadRequestException('Unit is not available for booking');
            }
            const booking = await prisma.booking.create({
                data: {
                    ...data,
                    status: shared_1.BookingStatus.INITIATED,
                },
            });
            await prisma.unit.update({
                where: { id: unit.id },
                data: { status: shared_1.UnitStatus.BOOKED },
            });
            const lead = await prisma.lead.findUnique({ where: { id: data.leadId } });
            if (lead?.brokerId) {
                const broker = await prisma.broker.findUnique({ where: { id: lead.brokerId } });
                if (broker) {
                    await prisma.commission.create({
                        data: {
                            brokerId: broker.id,
                            bookingId: booking.id,
                            amount: booking.bookingAmount * (broker.commissionPct / 100),
                        },
                    });
                }
            }
            return booking;
        });
    }
    async findAll(user, page = 1, limit = 50) {
        return this.prisma.booking.findMany({
            where: { lead: { tenantId: user.tenantId } },
            take: limit,
            skip: (page - 1) * limit,
            include: {
                unit: { include: { tower: { include: { project: true } } } },
                lead: true,
            },
            orderBy: { createdAt: 'desc' },
        });
    }
    async findOne(user, id) {
        const booking = await this.prisma.booking.findFirst({
            where: { id, lead: { tenantId: user.tenantId } },
            include: {
                payments: true,
                commissions: true,
                unit: true,
                lead: true,
                coBuyers: true,
                refunds: true,
            },
        });
        if (!booking)
            throw new common_1.NotFoundException('Booking not found');
        return booking;
    }
    async schedulePayments(user, id, payments) {
        const booking = await this.prisma.booking.findFirst({
            where: { id, unit: { tower: { project: { tenantId: user.tenantId } } } },
        });
        if (!booking)
            throw new common_1.NotFoundException('Booking not found');
        return this.prisma.$transaction(async (prisma) => {
            await prisma.payment.deleteMany({
                where: { bookingId: id, paidAt: null },
            });
            return prisma.payment.createMany({
                data: payments.map(p => ({
                    ...p,
                    bookingId: id,
                    dueDate: new Date(p.dueDate),
                })),
            });
        });
    }
    async recordPayment(user, id, paymentId, data) {
        const booking = await this.prisma.booking.findFirst({
            where: { id, unit: { tower: { project: { tenantId: user.tenantId } } } },
        });
        if (!booking)
            throw new common_1.NotFoundException('Booking not found');
        return this.prisma.payment.update({
            where: { id: paymentId, bookingId: id },
            data: {
                ...data,
                paidAt: data.paidAt ? new Date(data.paidAt) : new Date(),
            },
        });
    }
};
exports.BookingsService = BookingsService;
exports.BookingsService = BookingsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        audit_service_1.AuditService])
], BookingsService);
