// api/dist/apps/api/src/modules/brokers/brokers.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrokersService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const crypto = __importStar(require("crypto"));
let BrokersService = class BrokersService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async create(user, data) {
        const referralCode = data.referralCode || crypto.randomBytes(4).toString('hex').toUpperCase();
        return this.prisma.broker.create({
            data: {
                ...data,
                referralCode,
                tenantId: user.tenantId,
            },
        });
    }
    async findAll(user, page = 1, limit = 50) {
        return this.prisma.broker.findMany({
            where: { tenantId: user.tenantId },
            take: limit,
            skip: (page - 1) * limit,
            include: {
                _count: {
                    select: { leads: true, commissions: true }
                }
            },
            orderBy: { createdAt: 'desc' },
        });
    }
    async findOne(user, id) {
        const broker = await this.prisma.broker.findFirst({
            where: { id, tenantId: user.tenantId },
            include: { leads: true, commissions: true },
        });
        if (!broker)
            throw new common_1.NotFoundException('Broker not found');
        return broker;
    }
    async getStatement(user, id) {
        const broker = await this.prisma.broker.findFirst({
            where: { id, tenantId: user.tenantId },
            include: {
                commissions: {
                    include: {
                        booking: {
                            include: { unit: { include: { tower: { include: { project: true } } } } }
                        }
                    }
                }
            },
        });
        if (!broker)
            throw new common_1.NotFoundException('Broker not found');
        const totalCommission = broker.commissions.reduce((sum, c) => sum + c.amount, 0);
        const paidCommission = broker.commissions.filter(c => c.isPaid).reduce((sum, c) => sum + c.amount, 0);
        return {
            broker: {
                name: broker.name,
                email: broker.email,
                phone: broker.phone,
            },
            summary: {
                totalCommission,
                paidCommission,
                pendingCommission: totalCommission - paidCommission,
            },
            transactions: broker.commissions.map(c => ({
                id: c.id,
                amount: c.amount,
                isPaid: c.isPaid,
                paidAt: c.paidAt,
                createdAt: c.createdAt,
                project: c.booking.unit.tower.project.name,
                unit: c.booking.unit.unitNumber,
                buyer: c.booking.buyerName,
            })),
        };
    }
};
exports.BrokersService = BrokersService;
exports.BrokersService = BrokersService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], BrokersService);
