// api/dist/apps/api/src/modules/brokers/broker-disputes.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrokerDisputesController = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const jwt_auth_guard_1 = require("../../common/guards/jwt-auth.guard");
const roles_guard_1 = require("../../common/guards/roles.guard");
const roles_decorator_1 = require("../../common/decorators/roles.decorator");
const user_decorator_1 = require("../../common/decorators/user.decorator");
const shared_1 = require("@waterting/shared");
let BrokerDisputesController = class BrokerDisputesController {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async findAll(brokerId, user) {
        return this.prisma.brokerDispute.findMany({
            where: { brokerId, tenantId: user.tenantId },
            orderBy: { createdAt: 'desc' },
        });
    }
    async create(brokerId, user, dto) {
        return this.prisma.brokerDispute.create({
            data: {
                ...dto,
                brokerId,
                tenantId: user.tenantId,
                status: 'OPEN',
            },
        });
    }
    async resolve(id, dto) {
        return this.prisma.brokerDispute.update({
            where: { id },
            data: {
                resolution: dto.resolution,
                status: dto.status,
            },
        });
    }
};
exports.BrokerDisputesController = BrokerDisputesController;
__decorate([
    (0, common_1.Get)(),
    __param(0, (0, common_1.Param)('brokerId')),
    __param(1, (0, user_decorator_1.CurrentUser)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], BrokerDisputesController.prototype, "findAll", null);
__decorate([
    (0, common_1.Post)(),
    (0, roles_decorator_1.Roles)(shared_1.UserRole.TENANT_ADMIN, shared_1.UserRole.SALES_MANAGER),
    __param(0, (0, common_1.Param)('brokerId')),
    __param(1, (0, user_decorator_1.CurrentUser)()),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object, Object]),
    __metadata("design:returntype", Promise)
], BrokerDisputesController.prototype, "create", null);
__decorate([
    (0, common_1.Post)(':id/resolve'),
    (0, roles_decorator_1.Roles)(shared_1.UserRole.TENANT_ADMIN),
    __param(0, (0, common_1.Param)('id')),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], BrokerDisputesController.prototype, "resolve", null);
exports.BrokerDisputesController = BrokerDisputesController = __decorate([
    (0, common_1.Controller)('brokers/:brokerId/disputes'),
    (0, common_1.UseGuards)(jwt_auth_guard_1.JwtAuthGuard, roles_guard_1.RolesGuard),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], BrokerDisputesController);
