// api/dist/apps/api/src/modules/brokers/dispute.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrokerDisputeService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const audit_service_1 = require("../../common/audit/audit.service");
let BrokerDisputeService = class BrokerDisputeService {
    constructor(prisma, audit) {
        this.prisma = prisma;
        this.audit = audit;
    }
    async create(user, data) {
        const dispute = await this.prisma.brokerDispute.create({
            data: {
                ...data,
                tenantId: user.tenantId,
            },
        });
        await this.audit.log(user.tenantId, 'CREATE_DISPUTE', 'BrokerDispute', dispute.id, user.sub, null, dispute);
        return dispute;
    }
    async resolve(user, id, data) {
        const dispute = await this.prisma.brokerDispute.findUnique({ where: { id } });
        if (!dispute)
            throw new common_1.NotFoundException('Dispute not found');
        const updated = await this.prisma.brokerDispute.update({
            where: { id },
            data: {
                resolution: data.resolution,
                status: data.status,
                updatedAt: new Date(),
            },
        });
        await this.audit.log(user.tenantId, 'RESOLVE_DISPUTE', 'BrokerDispute', id, user.sub, dispute, updated);
        return updated;
    }
    async findAll(user) {
        return this.prisma.brokerDispute.findMany({
            where: { tenantId: user.tenantId },
            orderBy: { createdAt: 'desc' },
        });
    }
};
exports.BrokerDisputeService = BrokerDisputeService;
exports.BrokerDisputeService = BrokerDisputeService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        audit_service_1.AuditService])
], BrokerDisputeService);
