// api/dist/apps/api/src/modules/units/units.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnitsController = void 0;
const common_1 = require("@nestjs/common");
const units_service_1 = require("./units.service");
const jwt_auth_guard_1 = require("../../common/guards/jwt-auth.guard");
const roles_guard_1 = require("../../common/guards/roles.guard");
const roles_decorator_1 = require("../../common/decorators/roles.decorator");
const user_decorator_1 = require("../../common/decorators/user.decorator");
const shared_1 = require("@waterting/shared");
const prisma_service_1 = require("../../common/prisma/prisma.service");
let UnitsController = class UnitsController {
    constructor(unitsService, prisma) {
        this.unitsService = unitsService;
        this.prisma = prisma;
    }
    create(user, towerId, createUnitDto) {
        return this.unitsService.create(user, towerId, createUnitDto);
    }
    findAll(user, towerId) {
        return this.unitsService.findAllByTower(user, towerId);
    }
    async bulkPrice(user, dto) {
        return this.prisma.unit.updateMany({
            where: {
                towerId: dto.towerId,
                tower: { project: { tenantId: user.tenantId } },
                ...(dto.floor !== undefined && { floor: dto.floor }),
            },
            data: { basePrice: dto.basePrice, totalPrice: dto.basePrice },
        });
    }
    async holdUnit(user, id, dto) {
        const holdUntil = new Date(Date.now() + (dto.holdHours || 48) * 3600000);
        return this.prisma.unit.update({
            where: { id, tower: { project: { tenantId: user.tenantId } } },
            data: { status: shared_1.UnitStatus.RESERVED, holdUntil },
        });
    }
    updateStatus(user, id, status, holdUntil) {
        return this.unitsService.updateStatus(user, id, status, holdUntil ? new Date(holdUntil) : undefined);
    }
    async exportUnits(user, projectId, res) {
        const units = await this.prisma.unit.findMany({
            where: { tower: { projectId, project: { tenantId: user.tenantId } } },
            include: { tower: true },
        });
        let csv = 'Unit Number,Tower,Floor,Type,Carpet Area,Total Price,Status\n';
        units.forEach(u => {
            csv += `${u.unitNumber},${u.tower.name},${u.floor},${u.type},${u.carpetArea},${u.totalPrice},${u.status}\n`;
        });
        res.setHeader('Content-Type', 'text/csv');
        res.setHeader('Content-Disposition', `attachment; filename=inventory-export-${projectId}.csv`);
        res.send(csv);
    }
};
exports.UnitsController = UnitsController;
__decorate([
    (0, common_1.Post)('towers/:towerId/units'),
    (0, roles_decorator_1.Roles)(shared_1.UserRole.TENANT_ADMIN, shared_1.UserRole.SALES_MANAGER),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('towerId')),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, Object]),
    __metadata("design:returntype", void 0)
], UnitsController.prototype, "create", null);
__decorate([
    (0, common_1.Get)('towers/:towerId/units'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('towerId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String]),
    __metadata("design:returntype", void 0)
], UnitsController.prototype, "findAll", null);
__decorate([
    (0, common_1.Patch)('units/bulk-price'),
    (0, roles_decorator_1.Roles)(shared_1.UserRole.TENANT_ADMIN, shared_1.UserRole.SALES_MANAGER),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], UnitsController.prototype, "bulkPrice", null);
__decorate([
    (0, common_1.Patch)('units/:id/hold'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, Object]),
    __metadata("design:returntype", Promise)
], UnitsController.prototype, "holdUnit", null);
__decorate([
    (0, common_1.Patch)('units/:id/status'),
    (0, roles_decorator_1.Roles)(shared_1.UserRole.TENANT_ADMIN, shared_1.UserRole.SALES_MANAGER, shared_1.UserRole.SALES_AGENT),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __param(2, (0, common_1.Body)('status')),
    __param(3, (0, common_1.Body)('holdUntil')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, String, String]),
    __metadata("design:returntype", void 0)
], UnitsController.prototype, "updateStatus", null);
__decorate([
    (0, common_1.Get)('projects/:projectId/units/export'),
    (0, roles_decorator_1.Roles)(shared_1.UserRole.TENANT_ADMIN, shared_1.UserRole.SALES_MANAGER),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('projectId')),
    __param(2, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, Object]),
    __metadata("design:returntype", Promise)
], UnitsController.prototype, "exportUnits", null);
exports.UnitsController = UnitsController = __decorate([
    (0, common_1.Controller)(),
    (0, common_1.UseGuards)(jwt_auth_guard_1.JwtAuthGuard, roles_guard_1.RolesGuard),
    __metadata("design:paramtypes", [units_service_1.UnitsService,
        prisma_service_1.PrismaService])
], UnitsController);
