// api/dist/apps/api/src/modules/units/units.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnitsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const shared_1 = require("@waterting/shared");
const ai_follow_up_worker_1 = require("../../workers/ai-follow-up.worker");
let UnitsService = class UnitsService {
    constructor(prisma, aiFollowUp) {
        this.prisma = prisma;
        this.aiFollowUp = aiFollowUp;
    }
    async create(user, towerId, data) {
        const tower = await this.prisma.tower.findUnique({
            where: { id: towerId },
            include: { project: true },
        });
        if (!tower || tower.project.tenantId !== user.tenantId) {
            throw new common_1.NotFoundException('Tower not found');
        }
        return this.prisma.unit.create({
            data: {
                ...data,
                towerId,
            },
        });
    }
    async findAllByTower(user, towerId) {
        return this.prisma.unit.findMany({
            where: { towerId, tower: { project: { tenantId: user.tenantId } } },
        });
    }
    async updateStatus(user, id, status, holdUntil) {
        const unit = await this.prisma.unit.findFirst({
            where: { id, tower: { project: { tenantId: user.tenantId } } },
            include: { tower: { include: { project: true } } },
        });
        if (!unit)
            throw new common_1.NotFoundException('Unit not found');
        const updated = await this.prisma.unit.update({
            where: { id },
            data: {
                status,
                holdUntil: status === shared_1.UnitStatus.RESERVED ? holdUntil : null,
            },
        });
        if (status === shared_1.UnitStatus.AVAILABLE) {
            const availableCount = await this.prisma.unit.count({
                where: { tower: { projectId: unit.tower.projectId }, status: shared_1.UnitStatus.AVAILABLE }
            });
            await this.aiFollowUp.sendInventoryUrgency(unit.tower.projectId, availableCount);
        }
        return updated;
    }
    async updatePrice(user, id, newPrice) {
        const unit = await this.prisma.unit.findFirst({
            where: { id, tower: { project: { tenantId: user.tenantId } } },
            include: { tower: true },
        });
        if (!unit)
            throw new common_1.NotFoundException('Unit not found');
        const oldPrice = unit.basePrice;
        const updated = await this.prisma.unit.update({
            where: { id },
            data: { basePrice: newPrice },
        });
        if (newPrice < oldPrice) {
            await this.aiFollowUp.sendPriceDropAlert(unit.tower.projectId, newPrice);
        }
        return updated;
    }
};
exports.UnitsService = UnitsService;
exports.UnitsService = UnitsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        ai_follow_up_worker_1.AiFollowUpWorker])
], UnitsService);
