// api/dist/apps/api/src/modules/listings/listings.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListingsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
let ListingsService = class ListingsService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async create(user, data) {
        return this.prisma.listing.create({
            data: {
                ...data,
                tenantId: user.tenantId,
            },
        });
    }
    async findAll(user) {
        return this.prisma.listing.findMany({
            where: { tenantId: user.tenantId },
            include: { project: true },
        });
    }
    async syncToPlatform(listingId) {
        const listing = await this.prisma.listing.findUnique({
            where: { id: listingId },
        });
        if (!listing)
            throw new common_1.NotFoundException('Listing not found');
        await new Promise(resolve => setTimeout(resolve, 1000));
        return this.prisma.listing.update({
            where: { id: listingId },
            data: { platformId: `EXT-${Math.random().toString(36).substr(2, 9)}`, status: 'ACTIVE' },
        });
    }
    async handleWebhook(tenantId, platform, payload) {
        const leadData = {
            tenantId,
            name: payload.name || 'Platform Lead',
            phone: payload.phone,
            email: payload.email,
            source: platform,
            stage: 'NEW_LEAD',
        };
        return this.prisma.lead.upsert({
            where: {
                tenantId_phone: { tenantId, phone: leadData.phone }
            },
            update: { lastActivityAt: new Date() },
            create: leadData,
        });
    }
    async bulkUpdatePrices(tenantId, projectId, increasePct) {
        const factor = 1 + (increasePct / 100);
        await this.prisma.unit.updateMany({
            where: { tower: { projectId }, status: 'AVAILABLE' },
            data: {
                basePrice: { multiply: factor },
                totalPrice: { multiply: factor },
            },
        });
        return this.prisma.listing.updateMany({
            where: { tenantId, projectId },
            data: {
                price: { multiply: factor },
            },
        });
    }
};
exports.ListingsService = ListingsService;
exports.ListingsService = ListingsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], ListingsService);
