// api/dist/apps/api/src/modules/webhooks/webhooks.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhooksController = void 0;
const common_1 = require("@nestjs/common");
const webhooks_service_1 = require("./webhooks.service");
const leads_service_1 = require("../leads/leads.service");
const public_decorator_1 = require("../../common/decorators/public.decorator");
let WebhooksController = class WebhooksController {
    constructor(webhooksService, leadsService) {
        this.webhooksService = webhooksService;
        this.leadsService = leadsService;
    }
    verifyFacebook(query) {
        return this.webhooksService.verifyFacebookWebhook(query);
    }
    handleFacebookLead(tenantId, body) {
        return this.webhooksService.handleFacebookLead(tenantId, body);
    }
    async ninetyNineAcres(tenantId, body) {
        const lead = this.mapPortalLead(body, 'PORTAL_99ACRES');
        return this.leadsService.createFromWebhook({ ...lead, tenantId });
    }
    async magicBricks(tenantId, body) {
        const lead = this.mapPortalLead(body, 'PORTAL_MAGICBRICKS');
        return this.leadsService.createFromWebhook({ ...lead, tenantId });
    }
    mapPortalLead(body, source) {
        return {
            name: body.name ?? body.full_name ?? body.contact_name ?? 'Unknown',
            phone: body.phone ?? body.mobile ?? body.contact_number ?? '',
            email: body.email ?? body.email_id ?? undefined,
            source,
            utmSource: source,
        };
    }
};
exports.WebhooksController = WebhooksController;
__decorate([
    (0, public_decorator_1.Public)(),
    (0, common_1.Get)('facebook'),
    __param(0, (0, common_1.Query)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], WebhooksController.prototype, "verifyFacebook", null);
__decorate([
    (0, public_decorator_1.Public)(),
    (0, common_1.Post)('facebook/:tenantId'),
    __param(0, (0, common_1.Param)('tenantId')),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", void 0)
], WebhooksController.prototype, "handleFacebookLead", null);
__decorate([
    (0, public_decorator_1.Public)(),
    (0, common_1.Post)('99acres/:tenantId'),
    __param(0, (0, common_1.Param)('tenantId')),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], WebhooksController.prototype, "ninetyNineAcres", null);
__decorate([
    (0, public_decorator_1.Public)(),
    (0, common_1.Post)('magicbricks/:tenantId'),
    __param(0, (0, common_1.Param)('tenantId')),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], WebhooksController.prototype, "magicBricks", null);
exports.WebhooksController = WebhooksController = __decorate([
    (0, common_1.Controller)('webhooks'),
    __metadata("design:paramtypes", [webhooks_service_1.WebhooksService,
        leads_service_1.LeadsService])
], WebhooksController);
