// api/dist/apps/api/src/modules/webhooks/webhooks.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhooksService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const leads_service_1 = require("../leads/leads.service");
const shared_1 = require("@waterting/shared");
let WebhooksService = class WebhooksService {
    constructor(prisma, leadsService) {
        this.prisma = prisma;
        this.leadsService = leadsService;
    }
    verifyFacebookWebhook(query) {
        const mode = query['hub.mode'];
        const token = query['hub.verify_token'];
        const challenge = query['hub.challenge'];
        if (mode && token) {
            if (mode === 'subscribe' && token === process.env.FB_VERIFY_TOKEN) {
                return challenge;
            }
            else {
                throw new common_1.BadRequestException('Forbidden');
            }
        }
        throw new common_1.BadRequestException('Missing token');
    }
    async handleFacebookLead(tenantId, body) {
        body.entry?.forEach((entry) => {
            entry.changes?.forEach(async (change) => {
                if (change.field === 'leadgen') {
                    const leadgenId = change.value.leadgen_id;
                    const mappedLead = {
                        name: 'Facebook Lead',
                        phone: '9999999999',
                        email: '',
                        source: shared_1.LeadSource.FACEBOOK,
                        utmCampaign: leadgenId,
                    };
                    const superAdmin = await this.prisma.user.findFirst({ where: { tenantId } });
                    if (superAdmin) {
                        const mockUser = { sub: superAdmin.id, tenantId, role: superAdmin.role, email: superAdmin.email };
                        await this.leadsService.create(mockUser, mappedLead);
                    }
                }
            });
        });
        return { status: 'success' };
    }
};
exports.WebhooksService = WebhooksService;
exports.WebhooksService = WebhooksService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        leads_service_1.LeadsService])
], WebhooksService);
