// api/dist/apps/api/src/common/ai/ai.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var AIService_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AIService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../prisma/prisma.service");
const generative_ai_1 = require("@google/generative-ai");
let AIService = AIService_1 = class AIService {
    constructor(prisma) {
        this.prisma = prisma;
        this.logger = new common_1.Logger(AIService_1.name);
        this.gemini = new generative_ai_1.GoogleGenerativeAI(process.env.GEMINI_API_KEY || '');
    }
    async callOpenAI(endpoint, body) {
        const res = await fetch(`https://api.openai.com/v1/${endpoint}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`,
            },
            body: JSON.stringify(body),
        });
        if (!res.ok) {
            const error = await res.json();
            this.logger.error(`OpenAI error: ${JSON.stringify(error)}`);
            throw new common_1.InternalServerErrorException('AI Service Error');
        }
        return res.json();
    }
    async generateJSON(prompt) {
        if (process.env.GEMINI_API_KEY) {
            const model = this.gemini.getGenerativeModel({ model: 'gemini-1.5-flash' });
            const result = await model.generateContent(prompt + '\nReturn ONLY a valid JSON object.');
            const text = result.response.text();
            const jsonStr = text.substring(text.indexOf('{'), text.lastIndexOf('}') + 1);
            return JSON.parse(jsonStr);
        }
        const data = await this.callOpenAI('chat/completions', {
            model: 'gpt-4o',
            messages: [{ role: 'user', content: prompt }],
            response_format: { type: 'json_object' },
        });
        return JSON.parse(data.choices[0].message.content);
    }
    async generateText(prompt) {
        if (process.env.GEMINI_API_KEY) {
            const model = this.gemini.getGenerativeModel({ model: 'gemini-1.5-flash' });
            const result = await model.generateContent(prompt);
            return result.response.text();
        }
        const data = await this.callOpenAI('chat/completions', {
            model: 'gpt-4o',
            messages: [{ role: 'user', content: prompt }],
        });
        return data.choices[0].message.content;
    }
    async generateEmbedding(text) {
        if (process.env.GEMINI_API_KEY) {
            const model = this.gemini.getGenerativeModel({ model: 'text-embedding-004' });
            const result = await model.embedContent(text);
            return result.embedding.values;
        }
        const data = await this.callOpenAI('embeddings', {
            model: 'text-embedding-3-small',
            input: text,
            dimensions: 1536,
        });
        return data.data[0].embedding;
    }
    async upsertVector(id, tenantId, values, metadata, leadId, unitId) {
        try {
            const vector = `[${values.join(',')}]`;
            const existing = await this.prisma.embedding.findFirst({
                where: { id },
            });
            if (existing) {
                await this.prisma.$executeRaw `
          UPDATE "Embedding"
          SET "embedding" = ${vector}::vector, "metadata" = ${metadata}::jsonb, "updatedAt" = NOW()
          WHERE "id" = ${id}
        `;
            }
            else {
                await this.prisma.$executeRaw `
          INSERT INTO "Embedding" ("id", "tenantId", "leadId", "unitId", "embedding", "metadata", "createdAt")
          VALUES (${id}, ${tenantId}, ${leadId}, ${unitId}, ${vector}::vector, ${metadata}::jsonb, NOW())
        `;
            }
        }
        catch (error) {
            this.logger.error(`pgvector upsert failed for tenant ${tenantId}`, error);
            throw new common_1.InternalServerErrorException('Vector storage failed');
        }
    }
    async queryVector(tenantId, values, topK = 5) {
        try {
            const vector = `[${values.join(',')}]`;
            const results = await this.prisma.$queryRaw `
        SELECT "id", "leadId", "unitId", "metadata", "embedding" <=> ${vector}::vector AS distance
        FROM "Embedding"
        WHERE "tenantId" = ${tenantId}
        ORDER BY distance ASC
        LIMIT ${topK}
      `;
            return results;
        }
        catch (error) {
            this.logger.error(`pgvector query failed for tenant ${tenantId}`, error);
            throw new common_1.InternalServerErrorException('Vector search failed');
        }
    }
};
exports.AIService = AIService;
exports.AIService = AIService = AIService_1 = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], AIService);
