// api/dist/apps/api/src/common/ai/briefing.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var AiBriefingService_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AiBriefingService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../prisma/prisma.service");
const ai_service_1 = require("./ai.service");
let AiBriefingService = AiBriefingService_1 = class AiBriefingService {
    constructor(prisma, ai) {
        this.prisma = prisma;
        this.ai = ai;
        this.logger = new common_1.Logger(AiBriefingService_1.name);
    }
    async getBriefing(leadId) {
        const lead = await this.prisma.lead.findUnique({
            where: { id: leadId },
            include: {
                activities: { take: 10, orderBy: { createdAt: 'desc' } },
                siteVisits: true,
                project: true
            },
        });
        if (!lead)
            throw new Error('Lead not found');
        const activitySummary = lead.activities.map(a => `- ${a.title}: ${a.description || ''}`).join('\n');
        const visitSummary = lead.siteVisits.map(v => `- Visit on ${v.scheduledAt}: ${v.outcome || 'No outcome'}`).join('\n');
        const prompt = `
Analyze this real estate lead history and generate a pre-call briefing for the sales agent.

Lead Name: ${lead.name}
Project: ${lead.project?.name || 'Unknown'}
Budget: ${lead.budgetMax || 'Unknown'}
Timeline: ${lead.timeline || 'Unknown'}

Recent Activities:
${activitySummary}

Site Visits:
${visitSummary}

Return valid JSON:
{
  "summary": "1-sentence context",
  "talkingPoints": ["point 1", "point 2", "point 3"],
  "predictedObjection": "the most likely hurdle",
  "recommendedAction": "e.g. invite for site visit"
}
`;
        return this.ai.generateJSON(prompt);
    }
};
exports.AiBriefingService = AiBriefingService;
exports.AiBriefingService = AiBriefingService = AiBriefingService_1 = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        ai_service_1.AIService])
], AiBriefingService);
