// api/dist/apps/api/src/common/ai/property-rec.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var PropertyRecService_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropertyRecService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../prisma/prisma.service");
const ai_service_1 = require("./ai.service");
let PropertyRecService = PropertyRecService_1 = class PropertyRecService {
    constructor(prisma, ai) {
        this.prisma = prisma;
        this.ai = ai;
        this.logger = new common_1.Logger(PropertyRecService_1.name);
    }
    async recommendUnits(tenantId, leadId, topK = 3) {
        const lead = await this.prisma.lead.findUnique({
            where: { id: leadId },
            include: { project: true },
        });
        if (!lead)
            throw new Error('Lead not found');
        const prefString = `
      Budget: ${lead.budgetMin || 0} - ${lead.budgetMax || 'Any'}
      BHK: ${lead.preferredBHK || 'Any'}
      Project: ${lead.project?.name || 'Any'}
      Notes: ${lead.notes || ''}
      Timeline: ${lead.timeline || ''}
    `.trim();
        const embedding = await this.ai.generateEmbedding(prefString);
        const results = await this.prisma.$queryRaw `
      SELECT e."unitId", u."unitNumber", u."type", u."totalPrice", u."floor", t."name" as "towerName", p."name" as "projectName",
             e."embedding" <=> ${`[${embedding.join(',')}]`}::vector AS distance
      FROM "Embedding" e
      JOIN "Unit" u ON e."unitId" = u.id
      JOIN "Tower" t ON u."towerId" = t.id
      JOIN "Project" p ON t."projectId" = p.id
      WHERE e."tenantId" = ${tenantId} AND u.status = 'AVAILABLE' AND e."unitId" IS NOT NULL
      ORDER BY distance ASC
      LIMIT ${topK}
    `;
        if (results.length > 0) {
            const topMatch = results[0];
            const explanation = await this.ai.generateText(`
        Explain why this unit is a good match for this lead.
        Lead Preferences: ${prefString}
        Matched Unit: ${topMatch.type} in ${topMatch.projectName}, ${topMatch.towerName} Floor ${topMatch.floor}. Price: ${topMatch.totalPrice}.
        Keep it professional and concise (2 sentences).
      `);
            topMatch.explanation = explanation;
        }
        return results;
    }
    async syncUnitEmbedding(unitId) {
        const unit = await this.prisma.unit.findUnique({
            where: { id: unitId },
            include: { tower: { include: { project: true } } },
        });
        if (!unit)
            return;
        const unitString = `
      Project: ${unit.tower.project.name}
      Type: ${unit.type}
      Price: ${unit.totalPrice}
      Area: ${unit.carpetArea} sqft
      Floor: ${unit.floor}
      Facing: ${unit.facing || 'Unknown'}
      Amenities: ${unit.tower.project.amenities.join(', ')}
    `.trim();
        const embedding = await this.ai.generateEmbedding(unitString);
        await this.ai.upsertVector(`unit-${unit.id}`, unit.tower.project.tenantId, embedding, { unitId: unit.id, type: unit.type }, undefined, unit.id);
    }
};
exports.PropertyRecService = PropertyRecService;
exports.PropertyRecService = PropertyRecService = PropertyRecService_1 = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        ai_service_1.AIService])
], PropertyRecService);
