// api/dist/apps/api/src/modules/payments/payments.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const audit_service_1 = require("../../common/audit/audit.service");
let PaymentsService = class PaymentsService {
    constructor(prisma, audit) {
        this.prisma = prisma;
        this.audit = audit;
    }
    async create(user, bookingId, data) {
        const booking = await this.prisma.booking.findFirst({
            where: { id: bookingId, lead: { tenantId: user.tenantId } },
        });
        if (!booking)
            throw new common_1.NotFoundException('Booking not found');
        return this.prisma.payment.create({
            data: {
                ...data,
                bookingId,
            },
        });
    }
    async findAllByBooking(user, bookingId) {
        return this.prisma.payment.findMany({
            where: { bookingId, booking: { lead: { tenantId: user.tenantId } } },
            orderBy: { dueDate: 'asc' },
        });
    }
    async recordPayment(user, id, data) {
        const payment = await this.prisma.payment.findFirst({
            where: { id, booking: { lead: { tenantId: user.tenantId } } },
        });
        if (!payment)
            throw new common_1.NotFoundException('Payment not found');
        const updated = await this.prisma.payment.update({
            where: { id },
            data: {
                paidAt: data.paidAt ? new Date(data.paidAt) : new Date(),
                method: data.method,
                referenceNumber: data.referenceNumber,
                receiptUrl: data.receiptUrl,
                isVerified: false,
            },
        });
        await this.audit.log(user.tenantId, 'RECORD_PAYMENT', 'Payment', id, user.sub, payment, updated);
        return updated;
    }
    async verifyPayment(user, id, isVerified) {
        if (user.role !== 'ACCOUNTS' && user.role !== 'TENANT_ADMIN') {
            throw new Error('Only accounts role can verify payments');
        }
        const payment = await this.prisma.payment.findFirst({
            where: { id, booking: { lead: { tenantId: user.tenantId } } },
        });
        if (!payment)
            throw new common_1.NotFoundException('Payment not found');
        const updated = await this.prisma.payment.update({
            where: { id },
            data: { isVerified },
        });
        await this.audit.log(user.tenantId, 'VERIFY_PAYMENT', 'Payment', id, user.sub, payment, updated);
        return updated;
    }
};
exports.PaymentsService = PaymentsService;
exports.PaymentsService = PaymentsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        audit_service_1.AuditService])
], PaymentsService);
