// api/dist/apps/api/src/workers/ai-follow-up.worker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var AiFollowUpWorker_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AiFollowUpWorker = void 0;
const common_1 = require("@nestjs/common");
const schedule_1 = require("@nestjs/schedule");
const bull_1 = require("@nestjs/bull");
const prisma_service_1 = require("../common/prisma/prisma.service");
const ai_service_1 = require("../common/ai/ai.service");
let AiFollowUpWorker = AiFollowUpWorker_1 = class AiFollowUpWorker {
    constructor(prisma, ai, emailQueue) {
        this.prisma = prisma;
        this.ai = ai;
        this.emailQueue = emailQueue;
        this.logger = new common_1.Logger(AiFollowUpWorker_1.name);
    }
    async processSequences() {
        this.logger.log('Starting daily AI follow-up sequences...');
        const now = new Date();
        const noResponse = await this.prisma.lead.findMany({
            where: {
                emailOptOut: false,
                email: { not: null },
                stage: 'NEW_LEAD',
                createdAt: { lt: new Date(now.getTime() - 24 * 3600000) },
                activities: { none: { type: { in: ['CALL', 'EMAIL', 'WHATSAPP'] } } },
            },
            include: { project: true, assignedTo: true },
        });
        for (const lead of noResponse) {
            const daysSince = Math.floor((now.getTime() - lead.createdAt.getTime()) / 86400000);
            const sequenceDay = daysSince === 1 ? 1 : daysSince === 3 ? 3 : daysSince === 7 ? 7 : null;
            if (!sequenceDay)
                continue;
            await this.sendFollowUpEmail(lead, 'no-response-drip', sequenceDay);
        }
        const postVisit = await this.prisma.siteVisit.findMany({
            where: {
                outcome: 'INTERESTED',
                scheduledAt: { lt: new Date(now.getTime() - 4 * 3600000) },
            },
            include: { lead: { include: { project: true } } },
        });
        for (const visit of postVisit) {
            if (!visit.lead.email || visit.lead.emailOptOut)
                continue;
            if (visit.lead.stage === 'BOOKING_DONE')
                continue;
            await this.sendFollowUpEmail(visit.lead, 'post-visit-nurture', 0);
        }
        const activeLostDays = [90, 120, 180];
        for (const d of activeLostDays) {
            const targetDate = new Date(now);
            targetDate.setDate(targetDate.getDate() - d);
            const startOfDay = new Date(targetDate.setHours(0, 0, 0, 0));
            const endOfDay = new Date(targetDate.setHours(23, 59, 59, 999));
            const lost = await this.prisma.lead.findMany({
                where: {
                    stage: 'LOST',
                    emailOptOut: false,
                    email: { not: null },
                    updatedAt: { gte: startOfDay, lte: endOfDay }
                },
                include: { project: true }
            });
            for (const lead of lost) {
                await this.sendFollowUpEmail(lead, 're-engagement', d);
            }
        }
    }
    async sendPriceDropAlert(projectId, newPrice) {
        const hotLeads = await this.prisma.lead.findMany({
            where: {
                projectId,
                score: { gte: 61 },
                emailOptOut: false,
                email: { not: null },
                stage: { notIn: ['BOOKING_DONE', 'LOST'] },
            },
            include: { project: true },
        });
        for (const lead of hotLeads) {
            await this.sendFollowUpEmail(lead, 'price-drop-alert', 0, { newPrice });
        }
    }
    async sendInventoryUrgency(projectId, remainingUnits) {
        if (remainingUnits >= 5)
            return;
        const hotLeads = await this.prisma.lead.findMany({
            where: {
                projectId,
                score: { gte: 61 },
                emailOptOut: false,
                email: { not: null },
                stage: { notIn: ['BOOKING_DONE', 'LOST'] },
            },
            include: { project: true },
        });
        for (const lead of hotLeads) {
            await this.sendFollowUpEmail(lead, 'inventory-urgency', 0, { remainingUnits });
        }
    }
    async sendFollowUpEmail(lead, sequenceType, day, extra = {}) {
        const unsubUrl = `${process.env.FRONTEND_URL}/unsubscribe/${lead.id}`;
        const promptMap = {
            'no-response-drip': `Write a friendly, concise real estate follow-up email (Day ${day}). Lead name: ${lead.name}. Project: ${lead.project?.name ?? 'our project'}. Agent: ${lead.assignedTo?.name ?? 'our team'}. Tone: Warm, not pushy. 3 sentences max. End with a soft CTA like "Would you like to schedule a visit?" Return ONLY valid JSON: {"subject": "...", "body": "..."}`,
            'post-visit-nurture': `Write a post-site-visit follow-up email. Lead name: ${lead.name}. Project: ${lead.project?.name ?? 'our project'}. They visited and showed interest. Reinforce their decision. Tone: Excited, reassuring. 3-4 sentences. Return ONLY valid JSON: {"subject": "...", "body": "..."}`,
            'price-drop-alert': `Write a price drop alert email. Lead name: ${lead.name}. Project: ${lead.project?.name ?? 'our project'}. New price: ₹${extra.newPrice?.toLocaleString('en-IN') ?? 'reduced'}. Tone: Urgent but not spammy. 2-3 sentences. Return ONLY valid JSON: {"subject": "...", "body": "..."}`,
            'inventory-urgency': `Write a low inventory urgency email. Lead name: ${lead.name}. Project: ${lead.project?.name ?? 'our project'}. Only ${extra.remainingUnits} units remaining. Tone: Urgent, FOMO-driven but honest. 3 sentences max. Return ONLY valid JSON: {"subject": "...", "body": "..."}`,
            're-engagement': `Write a re-engagement email for a lead who dropped off ${day} days ago. Lead name: ${lead.name}. Project: ${lead.project?.name ?? 'our project'}. Tone: Casual, no hard sell. "Are you still looking?" energy. 2-3 sentences. Return ONLY valid JSON: {"subject": "...", "body": "..."}`,
        };
        const prompt = promptMap[sequenceType];
        if (!prompt)
            return;
        try {
            const email = await this.ai.generateJSON(prompt);
            const unsubLine = `<p style="font-size:11px;color:#888;margin-top:20px;"><a href="${unsubUrl}">Unsubscribe from email updates</a></p>`;
            await this.emailQueue.add('send', {
                to: lead.email,
                subject: email.subject,
                html: `<div style="font-family:sans-serif;max-width:600px;">${email.body.replace(/\n/g, '<br/>')}${unsubLine}</div>`,
            });
            await this.prisma.activity.create({
                data: {
                    leadId: lead.id,
                    type: 'AI_ACTION',
                    title: `AI email sent: ${sequenceType} (Day ${day})`,
                    description: `Subject: ${email.subject}`,
                },
            });
        }
        catch (err) {
            this.logger.error(`Failed to generate/send AI follow-up: ${err.message}`);
        }
    }
};
exports.AiFollowUpWorker = AiFollowUpWorker;
__decorate([
    (0, schedule_1.Cron)('0 6 * * *'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], AiFollowUpWorker.prototype, "processSequences", null);
exports.AiFollowUpWorker = AiFollowUpWorker = AiFollowUpWorker_1 = __decorate([
    (0, common_1.Injectable)(),
    __param(2, (0, bull_1.InjectQueue)('email')),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        ai_service_1.AIService, Object])
], AiFollowUpWorker);
