// api/dist/apps/api/src/workers/ai-lead-scoring.worker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AiLeadScoringWorker = void 0;
const bull_1 = require("@nestjs/bull");
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../common/prisma/prisma.service");
const events_gateway_1 = require("../gateways/events.gateway");
const ai_service_1 = require("../common/ai/ai.service");
let AiLeadScoringWorker = class AiLeadScoringWorker {
    constructor(prisma, events, ai) {
        this.prisma = prisma;
        this.events = events;
        this.ai = ai;
    }
    async scoreLead(job) {
        const { leadId, tenantId } = job.data;
        const lead = await this.prisma.lead.findUnique({
            where: { id: leadId },
            include: { project: true },
        });
        if (!lead)
            return;
        const totalBySource = await this.prisma.lead.count({
            where: { tenantId, source: lead.source },
        });
        const bookedBySource = await this.prisma.lead.count({
            where: { tenantId, source: lead.source, stage: 'BOOKING_DONE' },
        });
        const conversionRate = totalBySource > 0
            ? Math.round((bookedBySource / totalBySource) * 100)
            : 10;
        const units = await this.prisma.unit.findMany({
            where: { tower: { project: { id: lead.projectId ?? '' } } },
            select: { totalPrice: true },
        });
        const prices = units.map(u => u.totalPrice);
        const minPrice = prices.length ? Math.min(...prices) : 0;
        const maxPrice = prices.length ? Math.max(...prices) : 0;
        const prompt = `
Score this real estate lead from 0 to 100 for purchase conversion probability based on these PRD weights:

1. Budget match (25%): Lead budget up to ₹${lead.budgetMax ?? 'unknown'}, project price range ₹${minPrice}–₹${maxPrice}.
2. Source quality (20%): Historical conversion for ${lead.source} is ${conversionRate}%.
3. Response speed (15%): Lead was created ${Math.round((Date.now() - lead.createdAt.getTime()) / 3600000)} hours ago. (Lower is better)
4. Timeline urgency (15%): Wants to buy in ${lead.timeline ?? 'unknown'}. (Immediate is best)
5. Behavioral signals (15%): Source is ${lead.source}. (Organic/Portal is better than Mass Campaign)
6. Demographics (10%): No specific data yet.

PRD Score Labels:
- Cold: 0-30
- Warm: 31-60
- Hot: 61-80
- Very Hot: 81-100

Return ONLY valid JSON:
{"score": <number>, "label": "COLD|WARM|HOT|VERY_HOT", "reasoning": "string"}
`;
        const result = await this.ai.generateJSON(prompt);
        const score = Math.min(100, Math.max(0, result.score));
        const scoreLabel = result.label.toUpperCase().replace(' ', '_');
        await this.prisma.lead.update({
            where: { id: leadId },
            data: {
                score,
                scoreLabel: scoreLabel
            },
        });
        await this.prisma.activity.create({
            data: {
                leadId,
                type: 'AI_ACTION',
                title: `AI Score: ${score}/100 (${scoreLabel})`,
                description: result.reasoning,
            },
        });
        this.events.emitToTenant(tenantId, 'lead:scored', {
            leadId,
            score,
            scoreLabel,
        });
    }
};
exports.AiLeadScoringWorker = AiLeadScoringWorker;
__decorate([
    (0, bull_1.Process)('score-lead'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], AiLeadScoringWorker.prototype, "scoreLead", null);
exports.AiLeadScoringWorker = AiLeadScoringWorker = __decorate([
    (0, bull_1.Processor)('ai-scoring'),
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        events_gateway_1.EventsGateway,
        ai_service_1.AIService])
], AiLeadScoringWorker);
