// api/dist/apps/api/src/workers/payment-reminder.worker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentReminderWorker = void 0;
const common_1 = require("@nestjs/common");
const schedule_1 = require("@nestjs/schedule");
const bull_1 = require("@nestjs/bull");
const prisma_service_1 = require("../common/prisma/prisma.service");
let PaymentReminderWorker = class PaymentReminderWorker {
    constructor(prisma, emailQueue) {
        this.prisma = prisma;
        this.emailQueue = emailQueue;
    }
    async checkPayments() {
        const today = new Date();
        today.setHours(0, 0, 0, 0);
        const endToday = new Date(today);
        endToday.setHours(23, 59, 59, 999);
        const in7Days = new Date(today.getTime() + 7 * 24 * 60 * 60 * 1000);
        const endIn7Days = new Date(in7Days);
        endIn7Days.setHours(23, 59, 59, 999);
        const in1Day = new Date(today.getTime() + 1 * 24 * 60 * 60 * 1000);
        const endIn1Day = new Date(in1Day);
        endIn1Day.setHours(23, 59, 59, 999);
        const due7 = await this.prisma.payment.findMany({
            where: {
                paidAt: null,
                dueDate: { gte: in7Days, lte: endIn7Days },
            },
            include: { booking: { include: { lead: true } } },
        });
        for (const p of due7) {
            if (!p.booking.buyerEmail)
                continue;
            await this.emailQueue.add('send', {
                to: p.booking.buyerEmail,
                subject: `Payment due in 7 days — ₹${p.amount.toLocaleString('en-IN')}`,
                html: `Hi ${p.booking.buyerName}, your installment of ₹${p.amount.toLocaleString('en-IN')} is due on ${p.dueDate.toLocaleDateString('en-IN')}.`,
            });
        }
        const due1 = await this.prisma.payment.findMany({
            where: {
                paidAt: null,
                dueDate: { gte: in1Day, lte: endIn1Day },
            },
            include: { booking: { include: { lead: true } } },
        });
        for (const p of due1) {
            if (!p.booking.buyerEmail)
                continue;
            await this.emailQueue.add('send', {
                to: p.booking.buyerEmail,
                subject: `Payment due tomorrow — ₹${p.amount.toLocaleString('en-IN')}`,
                html: `Hi ${p.booking.buyerName}, your installment of ₹${p.amount.toLocaleString('en-IN')} is due tomorrow, ${p.dueDate.toLocaleDateString('en-IN')}.`,
            });
        }
        const overdue = await this.prisma.payment.findMany({
            where: { paidAt: null, dueDate: { lt: today }, isOverdue: false },
        });
        for (const p of overdue) {
            await this.prisma.payment.update({
                where: { id: p.id },
                data: { isOverdue: true },
            });
        }
    }
};
exports.PaymentReminderWorker = PaymentReminderWorker;
__decorate([
    (0, schedule_1.Cron)('0 9 * * *'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], PaymentReminderWorker.prototype, "checkPayments", null);
exports.PaymentReminderWorker = PaymentReminderWorker = __decorate([
    (0, common_1.Injectable)(),
    __param(1, (0, bull_1.InjectQueue)('email')),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService, Object])
], PaymentReminderWorker);
