// api/dist/apps/api/src/workers/pdf-generator.worker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PdfGeneratorWorker = void 0;
const bull_1 = require("@nestjs/bull");
const common_1 = require("@nestjs/common");
const pdfkit_1 = __importDefault(require("pdfkit"));
const client_s3_1 = require("@aws-sdk/client-s3");
const s3_request_presigner_1 = require("@aws-sdk/s3-request-presigner");
const prisma_service_1 = require("../common/prisma/prisma.service");
let PdfGeneratorWorker = class PdfGeneratorWorker {
    constructor(prisma) {
        this.prisma = prisma;
        this.r2 = new client_s3_1.S3Client({
            region: 'auto',
            endpoint: `https://${process.env.R2_ACCOUNT_ID}.r2.cloudflarestorage.com`,
            credentials: {
                accessKeyId: process.env.R2_ACCESS_KEY_ID,
                secretAccessKey: process.env.R2_SECRET_ACCESS_KEY,
            },
        });
    }
    async generateBookingConfirmation(job) {
        const booking = await this.prisma.booking.findUnique({
            where: { id: job.data.bookingId },
            include: {
                unit: { include: { tower: { include: { project: true } } } },
                payments: { take: 1 },
            },
        });
        if (!booking)
            return;
        const pdfBuffer = await new Promise((resolve) => {
            const doc = new pdfkit_1.default({ size: 'A4', margin: 50 });
            const chunks = [];
            doc.on('data', c => chunks.push(c));
            doc.on('end', () => resolve(Buffer.concat(chunks)));
            doc.fontSize(20).text('BOOKING CONFIRMATION', { align: 'center' });
            doc.moveDown();
            doc.fontSize(12).text(`Buyer: ${booking.buyerName}`);
            doc.text(`Project: ${booking.unit?.tower?.project?.name || 'N/A'}`);
            doc.text(`Unit: ${booking.unit?.unitNumber} — Floor ${booking.unit?.floor}`);
            doc.text(`BHK: ${booking.unit?.type}`);
            doc.text(`Area: ${booking.unit?.carpetArea} sq ft`);
            doc.text(`Booking Amount: ₹${booking.bookingAmount?.toLocaleString('en-IN')}`);
            doc.text(`Date: ${new Date().toLocaleDateString('en-IN')}`);
            doc.end();
        });
        const key = `${booking.unit?.tower?.project?.tenantId}/bookings/${booking.id}/confirmation.pdf`;
        await this.r2.send(new client_s3_1.PutObjectCommand({
            Bucket: process.env.R2_BUCKET_NAME,
            Key: key,
            Body: pdfBuffer,
            ContentType: 'application/pdf',
        }));
        const signedUrl = await (0, s3_request_presigner_1.getSignedUrl)(this.r2, new client_s3_1.GetObjectCommand({ Bucket: process.env.R2_BUCKET_NAME, Key: key }), { expiresIn: 3600 });
        await this.prisma.booking.update({
            where: { id: booking.id },
            data: { documents: { push: key } },
        });
        return { url: signedUrl };
    }
};
exports.PdfGeneratorWorker = PdfGeneratorWorker;
__decorate([
    (0, bull_1.Process)('booking-confirmation'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], PdfGeneratorWorker.prototype, "generateBookingConfirmation", null);
exports.PdfGeneratorWorker = PdfGeneratorWorker = __decorate([
    (0, bull_1.Processor)('pdf'),
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], PdfGeneratorWorker);
