// api/dist/apps/api/src/workers/stale-lead.worker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaleLeadWorker = void 0;
const common_1 = require("@nestjs/common");
const schedule_1 = require("@nestjs/schedule");
const prisma_service_1 = require("../common/prisma/prisma.service");
const events_gateway_1 = require("../gateways/events.gateway");
let StaleLeadWorker = class StaleLeadWorker {
    constructor(prisma, events) {
        this.prisma = prisma;
        this.events = events;
    }
    async checkStaleLeads() {
        const sevenDaysAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
        const threeDaysAgo = new Date(Date.now() - 3 * 24 * 60 * 60 * 1000);
        const twoDaysAgo = new Date(Date.now() - 2 * 24 * 60 * 60 * 1000);
        const stale = await this.prisma.lead.findMany({
            where: {
                stage: { notIn: ['BOOKING_DONE', 'LOST'] },
                lastActivityAt: { lt: sevenDaysAgo },
            },
            include: { assignedTo: true },
        });
        const hotStale = await this.prisma.lead.findMany({
            where: {
                stage: { notIn: ['BOOKING_DONE', 'LOST'] },
                score: { gte: 61 },
                lastActivityAt: { lt: threeDaysAgo },
            },
        });
        const contactedStale = await this.prisma.lead.findMany({
            where: {
                stage: 'CONTACTED',
                lastActivityAt: { lt: twoDaysAgo },
            },
        });
        const staleMap = new Map();
        [...stale, ...hotStale, ...contactedStale].forEach(lead => staleMap.set(lead.id, lead));
        const allStale = Array.from(staleMap.values());
        for (const lead of allStale) {
            const daysSince = Math.floor((Date.now() - (lead.lastActivityAt?.getTime() ?? lead.createdAt.getTime())) / 86400000);
            this.events.emitToTenant(lead.tenantId, 'notification', {
                title: `Stale Lead: ${lead.name}`,
                message: `No activity for ${daysSince} days — currently in ${lead.stage}`,
                type: 'STALE_LEAD',
                leadId: lead.id,
            });
        }
    }
};
exports.StaleLeadWorker = StaleLeadWorker;
__decorate([
    (0, schedule_1.Cron)('*/5 * * * *'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], StaleLeadWorker.prototype, "checkStaleLeads", null);
exports.StaleLeadWorker = StaleLeadWorker = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        events_gateway_1.EventsGateway])
], StaleLeadWorker);
