// api/dist/apps/api/src/workers/unit-hold-expiry.worker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnitHoldExpiryWorker = void 0;
const common_1 = require("@nestjs/common");
const schedule_1 = require("@nestjs/schedule");
const prisma_service_1 = require("../common/prisma/prisma.service");
const events_gateway_1 = require("../gateways/events.gateway");
let UnitHoldExpiryWorker = class UnitHoldExpiryWorker {
    constructor(prisma, events) {
        this.prisma = prisma;
        this.events = events;
    }
    async releaseExpiredHolds() {
        const expired = await this.prisma.unit.findMany({
            where: {
                status: 'RESERVED',
                holdUntil: { lt: new Date() },
            },
            include: { tower: { include: { project: true } } },
        });
        for (const unit of expired) {
            await this.prisma.unit.update({
                where: { id: unit.id },
                data: { status: 'AVAILABLE', holdUntil: null },
            });
            const tenantId = unit.tower?.project?.tenantId;
            if (tenantId) {
                this.events.emitToTenant(tenantId, 'unit:status', {
                    unitId: unit.id,
                    status: 'AVAILABLE',
                });
            }
        }
    }
};
exports.UnitHoldExpiryWorker = UnitHoldExpiryWorker;
__decorate([
    (0, schedule_1.Cron)('*/30 * * * *'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], UnitHoldExpiryWorker.prototype, "releaseExpiredHolds", null);
exports.UnitHoldExpiryWorker = UnitHoldExpiryWorker = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        events_gateway_1.EventsGateway])
], UnitHoldExpiryWorker);
