// api/dist/apps/api/src/workers/visit-reminder.worker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VisitReminderWorker = void 0;
const common_1 = require("@nestjs/common");
const schedule_1 = require("@nestjs/schedule");
const bull_1 = require("@nestjs/bull");
const prisma_service_1 = require("../common/prisma/prisma.service");
let VisitReminderWorker = class VisitReminderWorker {
    constructor(prisma, emailQueue) {
        this.prisma = prisma;
        this.emailQueue = emailQueue;
    }
    async send24hReminders() {
        const tomorrow = new Date();
        tomorrow.setDate(tomorrow.getDate() + 1);
        const start = new Date(tomorrow.setHours(0, 0, 0, 0));
        const end = new Date(tomorrow.setHours(23, 59, 59, 999));
        const visits = await this.prisma.siteVisit.findMany({
            where: { scheduledAt: { gte: start, lte: end }, outcome: null },
            include: { lead: true, agent: true },
        });
        for (const visit of visits) {
            if (!visit.lead.email || visit.lead.emailOptOut)
                continue;
            await this.emailQueue.add('send', {
                to: visit.lead.email,
                subject: `Reminder: Visit tomorrow`,
                html: `Hi ${visit.lead.name}, your site visit is tomorrow at ${visit.scheduledAt.toLocaleString('en-IN')}`,
            });
        }
    }
    async send2hReminders() {
        const now = new Date();
        const twoHoursLater = new Date(now.getTime() + 2 * 60 * 60 * 1000);
        const windowStart = new Date(twoHoursLater.getTime() - 10 * 60 * 1000);
        const windowEnd = new Date(twoHoursLater.getTime() + 10 * 60 * 1000);
        const visits = await this.prisma.siteVisit.findMany({
            where: { scheduledAt: { gte: windowStart, lte: windowEnd }, outcome: null },
            include: { lead: true },
        });
        for (const visit of visits) {
            if (!visit.lead.email || visit.lead.emailOptOut)
                continue;
            await this.emailQueue.add('send', {
                to: visit.lead.email,
                subject: `Your visit is in 2 hours`,
                html: `Hi ${visit.lead.name}, your site visit is in 2 hours at ${visit.scheduledAt.toLocaleTimeString('en-IN')}.`,
            });
        }
    }
};
exports.VisitReminderWorker = VisitReminderWorker;
__decorate([
    (0, schedule_1.Cron)('0 8 * * *'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], VisitReminderWorker.prototype, "send24hReminders", null);
__decorate([
    (0, schedule_1.Cron)('0 * * * *'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], VisitReminderWorker.prototype, "send2hReminders", null);
exports.VisitReminderWorker = VisitReminderWorker = __decorate([
    (0, common_1.Injectable)(),
    __param(1, (0, bull_1.InjectQueue)('email')),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService, Object])
], VisitReminderWorker);
