// api/dist/apps/api/src/modules/dashboard/dashboard.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DashboardService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
let DashboardService = class DashboardService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async getStats(user) {
        const tenantId = user.tenantId;
        const [totalLeads, newLeads, activeLeads, bookings, revenue, siteVisits, recentLeads, stageDistribution] = await Promise.all([
            this.prisma.lead.count({ where: { tenantId } }),
            this.prisma.lead.count({ where: { tenantId, stage: 'NEW_LEAD' } }),
            this.prisma.lead.count({ where: { tenantId, isArchived: false, stage: { notIn: ['LOST', 'BOOKING_DONE'] } } }),
            this.prisma.booking.count({ where: { lead: { tenantId } } }),
            this.prisma.payment.aggregate({
                where: { booking: { lead: { tenantId } }, paidAt: { not: null } },
                _sum: { amount: true },
            }),
            this.prisma.siteVisit.count({
                where: { lead: { tenantId }, scheduledAt: { gte: new Date(new Date().setHours(0, 0, 0, 0)) } },
            }),
            this.prisma.lead.findMany({
                where: { tenantId },
                take: 10,
                orderBy: { createdAt: 'desc' },
                include: { project: true, assignedTo: { select: { id: true, name: true } } },
            }),
            this.prisma.lead.groupBy({
                by: ['stage'],
                where: { tenantId },
                _count: { id: true },
            }),
        ]);
        return {
            totalLeads,
            newLeads,
            activeLeads,
            totalBookings: bookings,
            totalRevenue: revenue._sum.amount || 0,
            todaySiteVisits: siteVisits,
            recentLeads,
            stageDistribution: stageDistribution.map((s) => ({
                stage: s.stage,
                count: s._count.id,
            })),
        };
    }
};
exports.DashboardService = DashboardService;
exports.DashboardService = DashboardService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], DashboardService);
