// api/dist/apps/api/src/modules/automations/automations.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutomationsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
let AutomationsService = class AutomationsService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async create(user, data) {
        return this.prisma.automation.create({
            data: {
                ...data,
                tenantId: user.tenantId,
            },
        });
    }
    async findAll(user) {
        return this.prisma.automation.findMany({
            where: { tenantId: user.tenantId },
        });
    }
    async toggle(user, id) {
        const automation = await this.prisma.automation.findFirst({
            where: { id, tenantId: user.tenantId },
        });
        if (!automation)
            throw new common_1.NotFoundException('Automation not found');
        return this.prisma.automation.update({
            where: { id },
            data: { isEnabled: !automation.isEnabled },
        });
    }
};
exports.AutomationsService = AutomationsService;
exports.AutomationsService = AutomationsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], AutomationsService);
