// api/dist/packages/shared/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityType = exports.VisitOutcome = exports.BookingStatus = exports.PaymentMethod = exports.ProjectStatus = exports.ProjectType = exports.UnitType = exports.UnitStatus = exports.ScoreLabel = exports.LeadSource = exports.LeadStage = exports.UserRole = exports.Plan = void 0;
var Plan;
(function (Plan) {
    Plan["STARTER"] = "STARTER";
    Plan["GROWTH"] = "GROWTH";
    Plan["PROFESSIONAL"] = "PROFESSIONAL";
    Plan["ENTERPRISE"] = "ENTERPRISE";
})(Plan || (exports.Plan = Plan = {}));
var UserRole;
(function (UserRole) {
    UserRole["SUPER_ADMIN"] = "SUPER_ADMIN";
    UserRole["TENANT_ADMIN"] = "TENANT_ADMIN";
    UserRole["SALES_MANAGER"] = "SALES_MANAGER";
    UserRole["SALES_AGENT"] = "SALES_AGENT";
    UserRole["ACCOUNTS"] = "ACCOUNTS";
    UserRole["BROKER"] = "BROKER";
    UserRole["BUYER"] = "BUYER";
})(UserRole || (exports.UserRole = UserRole = {}));
var LeadStage;
(function (LeadStage) {
    LeadStage["NEW_LEAD"] = "NEW_LEAD";
    LeadStage["CONTACTED"] = "CONTACTED";
    LeadStage["INTERESTED"] = "INTERESTED";
    LeadStage["VISIT_SCHEDULED"] = "VISIT_SCHEDULED";
    LeadStage["VISIT_DONE"] = "VISIT_DONE";
    LeadStage["NEGOTIATION"] = "NEGOTIATION";
    LeadStage["BOOKING_DONE"] = "BOOKING_DONE";
    LeadStage["LOST"] = "LOST";
})(LeadStage || (exports.LeadStage = LeadStage = {}));
var LeadSource;
(function (LeadSource) {
    LeadSource["WEBSITE"] = "WEBSITE";
    LeadSource["LANDING_PAGE"] = "LANDING_PAGE";
    LeadSource["FACEBOOK"] = "FACEBOOK";
    LeadSource["GOOGLE"] = "GOOGLE";
    LeadSource["WHATSAPP"] = "WHATSAPP";
    LeadSource["PORTAL_99ACRES"] = "PORTAL_99ACRES";
    LeadSource["PORTAL_MAGICBRICKS"] = "PORTAL_MAGICBRICKS";
    LeadSource["PORTAL_HOUSING"] = "PORTAL_HOUSING";
    LeadSource["MANUAL"] = "MANUAL";
    LeadSource["BROKER"] = "BROKER";
    LeadSource["QR_CODE"] = "QR_CODE";
})(LeadSource || (exports.LeadSource = LeadSource = {}));
var ScoreLabel;
(function (ScoreLabel) {
    ScoreLabel["COLD"] = "COLD";
    ScoreLabel["WARM"] = "WARM";
    ScoreLabel["HOT"] = "HOT";
    ScoreLabel["VERY_HOT"] = "VERY_HOT";
})(ScoreLabel || (exports.ScoreLabel = ScoreLabel = {}));
var UnitStatus;
(function (UnitStatus) {
    UnitStatus["AVAILABLE"] = "AVAILABLE";
    UnitStatus["RESERVED"] = "RESERVED";
    UnitStatus["BOOKED"] = "BOOKED";
    UnitStatus["SOLD"] = "SOLD";
    UnitStatus["BLOCKED"] = "BLOCKED";
})(UnitStatus || (exports.UnitStatus = UnitStatus = {}));
var UnitType;
(function (UnitType) {
    UnitType["ONE_BHK"] = "ONE_BHK";
    UnitType["TWO_BHK"] = "TWO_BHK";
    UnitType["THREE_BHK"] = "THREE_BHK";
    UnitType["FOUR_BHK"] = "FOUR_BHK";
    UnitType["STUDIO"] = "STUDIO";
    UnitType["PENTHOUSE"] = "PENTHOUSE";
    UnitType["VILLA"] = "VILLA";
    UnitType["PLOT"] = "PLOT";
})(UnitType || (exports.UnitType = UnitType = {}));
var ProjectType;
(function (ProjectType) {
    ProjectType["RESIDENTIAL"] = "RESIDENTIAL";
    ProjectType["COMMERCIAL"] = "COMMERCIAL";
    ProjectType["MIXED"] = "MIXED";
    ProjectType["VILLA"] = "VILLA";
    ProjectType["TOWNSHIP"] = "TOWNSHIP";
})(ProjectType || (exports.ProjectType = ProjectType = {}));
var ProjectStatus;
(function (ProjectStatus) {
    ProjectStatus["ACTIVE"] = "ACTIVE";
    ProjectStatus["COMPLETED"] = "COMPLETED";
    ProjectStatus["UPCOMING"] = "UPCOMING";
    ProjectStatus["ON_HOLD"] = "ON_HOLD";
})(ProjectStatus || (exports.ProjectStatus = ProjectStatus = {}));
var PaymentMethod;
(function (PaymentMethod) {
    PaymentMethod["UPI"] = "UPI";
    PaymentMethod["NEFT"] = "NEFT";
    PaymentMethod["RTGS"] = "RTGS";
    PaymentMethod["CHEQUE"] = "CHEQUE";
    PaymentMethod["CASH"] = "CASH";
    PaymentMethod["DEMAND_DRAFT"] = "DEMAND_DRAFT";
})(PaymentMethod || (exports.PaymentMethod = PaymentMethod = {}));
var BookingStatus;
(function (BookingStatus) {
    BookingStatus["INITIATED"] = "INITIATED";
    BookingStatus["PAYMENT_PENDING"] = "PAYMENT_PENDING";
    BookingStatus["CONFIRMED"] = "CONFIRMED";
    BookingStatus["CANCELLED"] = "CANCELLED";
})(BookingStatus || (exports.BookingStatus = BookingStatus = {}));
var VisitOutcome;
(function (VisitOutcome) {
    VisitOutcome["INTERESTED"] = "INTERESTED";
    VisitOutcome["NOT_INTERESTED"] = "NOT_INTERESTED";
    VisitOutcome["NEED_MORE_TIME"] = "NEED_MORE_TIME";
    VisitOutcome["BOOKED"] = "BOOKED";
    VisitOutcome["NO_SHOW"] = "NO_SHOW";
})(VisitOutcome || (exports.VisitOutcome = VisitOutcome = {}));
var ActivityType;
(function (ActivityType) {
    ActivityType["CALL"] = "CALL";
    ActivityType["EMAIL"] = "EMAIL";
    ActivityType["WHATSAPP"] = "WHATSAPP";
    ActivityType["NOTE"] = "NOTE";
    ActivityType["STAGE_CHANGE"] = "STAGE_CHANGE";
    ActivityType["VISIT_SCHEDULED"] = "VISIT_SCHEDULED";
    ActivityType["VISIT_COMPLETED"] = "VISIT_COMPLETED";
    ActivityType["DOCUMENT_SHARED"] = "DOCUMENT_SHARED";
    ActivityType["AI_ACTION"] = "AI_ACTION";
})(ActivityType || (exports.ActivityType = ActivityType = {}));
