// api/dist/apps/api/src/modules/portal/portal-auth.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortalAuthService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const jwt_1 = require("@nestjs/jwt");
let PortalAuthService = class PortalAuthService {
    constructor(prisma, jwt) {
        this.prisma = prisma;
        this.jwt = jwt;
        this.otpStore = new Map();
    }
    async requestOTP(email) {
        const booking = await this.prisma.booking.findFirst({
            where: { buyerEmail: email },
        });
        if (!booking)
            throw new common_1.UnauthorizedException('Access denied. No booking found for this email.');
        const code = Math.floor(100000 + Math.random() * 900000).toString();
        this.otpStore.set(email, { code, expires: Date.now() + 600000 });
        console.log(`[SMTP MOCK] OTP for ${email}: ${code}`);
        return { message: 'OTP sent successfully' };
    }
    async verifyOTP(email, code) {
        const stored = this.otpStore.get(email);
        if (!stored || stored.code !== code || stored.expires < Date.now()) {
            throw new common_1.UnauthorizedException('Invalid or expired OTP');
        }
        this.otpStore.delete(email);
        const booking = await this.prisma.booking.findFirst({
            where: { buyerEmail: email },
            include: { lead: true },
        });
        const payload = {
            sub: booking?.leadId,
            email,
            role: 'BUYER',
            tenantId: booking?.lead.tenantId,
        };
        return {
            token: this.jwt.sign(payload),
            bookingId: booking?.id,
        };
    }
};
exports.PortalAuthService = PortalAuthService;
exports.PortalAuthService = PortalAuthService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        jwt_1.JwtService])
], PortalAuthService);
