// api/dist/apps/api/src/modules/portal/portal.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortalService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const jwt_1 = require("@nestjs/jwt");
const crypto = __importStar(require("crypto"));
let PortalService = class PortalService {
    constructor(prisma, jwtService) {
        this.prisma = prisma;
        this.jwtService = jwtService;
        this.otpStore = new Map();
    }
    async requestOtp(email) {
        const booking = await this.prisma.booking.findFirst({
            where: { buyerEmail: email },
        });
        if (!booking)
            throw new common_1.NotFoundException('No booking found for this email');
        const otp = crypto.randomInt(100000, 999999).toString();
        this.otpStore.set(email, { otp, expiresAt: Date.now() + 10 * 60 * 1000 });
        return { message: 'OTP sent to your email', ...(process.env.NODE_ENV !== 'production' ? { otp } : {}) };
    }
    async verifyOtp(email, otp) {
        const stored = this.otpStore.get(email);
        if (!stored || stored.otp !== otp || stored.expiresAt < Date.now()) {
            throw new common_1.UnauthorizedException('Invalid or expired OTP');
        }
        this.otpStore.delete(email);
        const token = this.jwtService.sign({ email, type: 'portal' });
        return { access_token: token };
    }
    async getDashboard(email) {
        const bookings = await this.prisma.booking.findMany({
            where: { buyerEmail: email },
            include: {
                unit: { include: { tower: { include: { project: true } } } },
                payments: { orderBy: { dueDate: 'asc' } },
            },
        });
        return { bookings };
    }
    async getPayments(email) {
        const bookings = await this.prisma.booking.findMany({
            where: { buyerEmail: email },
            include: { payments: { orderBy: { dueDate: 'asc' } } },
        });
        return bookings.flatMap((b) => b.payments);
    }
    async getDocuments(email) {
        const bookings = await this.prisma.booking.findMany({
            where: { buyerEmail: email },
            include: {
                lead: { include: { activities: { where: { type: 'DOCUMENT_SHARED' } } } }
            },
        });
        return bookings.map(b => ({
            bookingId: b.id,
            documents: b.lead?.activities || []
        }));
    }
    async getProperty(email) {
        const booking = await this.prisma.booking.findFirst({
            where: { buyerEmail: email },
            include: {
                unit: { include: { tower: { include: { project: true } } } }
            }
        });
        if (!booking)
            throw new common_1.NotFoundException();
        return booking.unit;
    }
};
exports.PortalService = PortalService;
exports.PortalService = PortalService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        jwt_1.JwtService])
], PortalService);
