// api/dist/apps/api/src/modules/leads/leads.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeadsController = void 0;
const common_1 = require("@nestjs/common");
const leads_service_1 = require("./leads.service");
const jwt_auth_guard_1 = require("../../common/guards/jwt-auth.guard");
const roles_guard_1 = require("../../common/guards/roles.guard");
const roles_decorator_1 = require("../../common/decorators/roles.decorator");
const user_decorator_1 = require("../../common/decorators/user.decorator");
const shared_1 = require("@waterting/shared");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const ai_service_1 = require("../../common/ai/ai.service");
let LeadsController = class LeadsController {
    constructor(leadsService, prisma, ai) {
        this.leadsService = leadsService;
        this.prisma = prisma;
        this.ai = ai;
    }
    create(user, createLeadDto) {
        return this.leadsService.create(user, createLeadDto);
    }
    findAll(user, page, limit) {
        return this.leadsService.findAll(user, Number(page || 1), Number(limit || 50));
    }
    findOne(user, id) {
        return this.leadsService.findOne(user, id);
    }
    update(user, id, data) {
        return this.leadsService.update(user, id, data);
    }
    assign(user, id, userId) {
        return this.leadsService.assign(user, id, userId);
    }
    addNote(user, id, title, description) {
        return this.leadsService.addNote(user, id, title, description);
    }
    addCall(user, id, duration, outcome) {
        return this.leadsService.addCall(user, id, duration, outcome);
    }
    updateStage(user, id, stage) {
        return this.leadsService.updateStage(user, id, stage);
    }
    async unsubscribe(id, req) {
        await this.prisma.lead.update({
            where: { id, tenantId: req.user.tenantId },
            data: { emailOptOut: true },
        });
        return { message: 'Unsubscribed' };
    }
    async publicUnsubscribe(leadId, res) {
        await this.prisma.lead.update({
            where: { id: leadId },
            data: { emailOptOut: true },
        });
        res.send('<h2>You have been unsubscribed.</h2><p>You will no longer receive emails from us.</p>');
    }
    async bulkAction(dto, req) {
        const { action, leadIds, payload } = dto;
        switch (action) {
            case 'reassign':
                await this.prisma.lead.updateMany({
                    where: { id: { in: leadIds }, tenantId: req.user.tenantId },
                    data: { assignedToId: payload.assignedToId },
                });
                break;
            case 'stage':
                for (const id of leadIds) {
                    await this.leadsService.updateStage(req.user, id, payload.stage);
                }
                break;
        }
        return { updated: leadIds.length };
    }
    async exportCSV(query, req, res) {
        const leads = await this.prisma.lead.findMany({
            where: { tenantId: req.user.tenantId },
            take: 10000,
            include: {
                project: true,
                assignedTo: true,
            }
        });
        const header = 'Name,Phone,Email,Source,Stage,Score,Project,Assigned To,Created\n';
        const rows = leads.map((l) => `"${l.name}","${l.phone}","${l.email ?? ''}","${l.source}","${l.stage}","${l.score ?? ''}","${l.project?.name ?? ''}","${l.assignedTo?.name ?? ''}","${l.createdAt}"`).join('\n');
        res.setHeader('Content-Type', 'text/csv');
        res.setHeader('Content-Disposition', 'attachment; filename=leads.csv');
        res.send(header + rows);
    }
    async getAIBrief(id, user) {
        const lead = await this.prisma.lead.findUnique({
            where: { id, tenantId: user.tenantId },
            include: {
                activities: { take: 5, orderBy: { createdAt: 'desc' } },
                project: true,
            },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        const prompt = `
Generate a 5-point pre-call sales brief for this real estate lead.
Lead: ${lead.name}, Budget: ₹${lead.budgetMax ?? 'unknown'}
AI Score: ${lead.score ?? 'not scored'}/100 (${lead.scoreLabel ?? 'unknown'})
Project: ${lead.project?.name ?? 'unknown'}
Recent activity: ${lead.activities.map(a => a.description).join('; ') || 'none'}

Return ONLY valid JSON:
{
  "background": "1 sentence",
  "scoreExplanation": "why this score",
  "likelyObjections": ["objection 1", "objection 2"],
  "recommendedUnit": "best unit based on budget",
  "suggestedOpener": "exact opening line for the call"
}`;
        return this.ai.generateJSON(prompt);
    }
    async getRecommendations(id, user) {
        const lead = await this.prisma.lead.findUnique({ where: { id, tenantId: user.tenantId } });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        return this.prisma.unit.findMany({
            where: {
                status: 'AVAILABLE',
                tower: { project: { tenantId: user.tenantId } },
                type: lead.preferredBHK ? lead.preferredBHK : undefined,
                totalPrice: { lte: lead.budgetMax || 100000000 }
            },
            take: 3,
            orderBy: { totalPrice: 'desc' }
        });
    }
    async getAISummary(id, user) {
        const lead = await this.prisma.lead.findUnique({
            where: { id, tenantId: user.tenantId },
            include: {
                activities: { take: 10, orderBy: { createdAt: 'desc' } },
                project: true,
            },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        const prompt = `Summarize this real estate lead's journey and intent in 3 concise bullet points.
    Name: ${lead.name}
    Current Stage: ${lead.stage}
    Budget: ₹${lead.budgetMin} - ₹${lead.budgetMax}
    Recent Actions: ${lead.activities.map(a => a.description).join('; ')}`;
        const summary = await this.ai.generateText(prompt);
        return { summary };
    }
};
exports.LeadsController = LeadsController;
__decorate([
    (0, common_1.Post)(),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "create", null);
__decorate([
    (0, common_1.Get)(),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Query)('page')),
    __param(2, (0, common_1.Query)('limit')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Number, Number]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "findAll", null);
__decorate([
    (0, common_1.Get)(':id'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "findOne", null);
__decorate([
    (0, common_1.Patch)(':id'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, Object]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "update", null);
__decorate([
    (0, common_1.Patch)(':id/assign'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __param(2, (0, common_1.Body)('userId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, String]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "assign", null);
__decorate([
    (0, common_1.Post)(':id/notes'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __param(2, (0, common_1.Body)('title')),
    __param(3, (0, common_1.Body)('description')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, String, String]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "addNote", null);
__decorate([
    (0, common_1.Post)(':id/calls'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __param(2, (0, common_1.Body)('duration')),
    __param(3, (0, common_1.Body)('outcome')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, String, String]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "addCall", null);
__decorate([
    (0, common_1.Patch)(':id/stage'),
    __param(0, (0, user_decorator_1.CurrentUser)()),
    __param(1, (0, common_1.Param)('id')),
    __param(2, (0, common_1.Body)('stage')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, String, String]),
    __metadata("design:returntype", void 0)
], LeadsController.prototype, "updateStage", null);
__decorate([
    (0, common_1.Patch)(':id/unsubscribe'),
    __param(0, (0, common_1.Param)('id')),
    __param(1, (0, common_1.Request)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], LeadsController.prototype, "unsubscribe", null);
__decorate([
    (0, common_1.Get)('unsubscribe/:leadId'),
    __param(0, (0, common_1.Param)('leadId')),
    __param(1, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], LeadsController.prototype, "publicUnsubscribe", null);
__decorate([
    (0, common_1.Post)('bulk'),
    (0, roles_decorator_1.Roles)(shared_1.UserRole.TENANT_ADMIN, shared_1.UserRole.SALES_MANAGER),
    __param(0, (0, common_1.Body)()),
    __param(1, (0, common_1.Request)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], LeadsController.prototype, "bulkAction", null);
__decorate([
    (0, common_1.Get)('export/csv'),
    __param(0, (0, common_1.Query)()),
    __param(1, (0, common_1.Request)()),
    __param(2, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], LeadsController.prototype, "exportCSV", null);
__decorate([
    (0, common_1.Get)(':id/ai-brief'),
    __param(0, (0, common_1.Param)('id')),
    __param(1, (0, user_decorator_1.CurrentUser)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], LeadsController.prototype, "getAIBrief", null);
__decorate([
    (0, common_1.Get)(':id/recommendations'),
    __param(0, (0, common_1.Param)('id')),
    __param(1, (0, user_decorator_1.CurrentUser)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], LeadsController.prototype, "getRecommendations", null);
__decorate([
    (0, common_1.Get)(':id/ai-summary'),
    __param(0, (0, common_1.Param)('id')),
    __param(1, (0, user_decorator_1.CurrentUser)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Promise)
], LeadsController.prototype, "getAISummary", null);
exports.LeadsController = LeadsController = __decorate([
    (0, common_1.Controller)('leads'),
    (0, common_1.UseGuards)(jwt_auth_guard_1.JwtAuthGuard, roles_guard_1.RolesGuard),
    __metadata("design:paramtypes", [leads_service_1.LeadsService,
        prisma_service_1.PrismaService,
        ai_service_1.AIService])
], LeadsController);
