// api/dist/apps/api/src/modules/leads/leads.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeadsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const shared_1 = require("@waterting/shared");
const bull_1 = require("@nestjs/bull");
const events_gateway_1 = require("../../gateways/events.gateway");
const notifications_service_1 = require("../notifications/notifications.service");
let LeadsService = class LeadsService {
    constructor(prisma, gateway, notificationsService, aiScoringQueue, emailQueue, pdfQueue) {
        this.prisma = prisma;
        this.gateway = gateway;
        this.notificationsService = notificationsService;
        this.aiScoringQueue = aiScoringQueue;
        this.emailQueue = emailQueue;
        this.pdfQueue = pdfQueue;
    }
    normalizePhone(phone) {
        return phone.replace(/\D/g, '');
    }
    async create(user, data) {
        const phone = this.normalizePhone(data.phone);
        const existing = await this.prisma.lead.findFirst({
            where: { tenantId: user.tenantId, phone },
        });
        if (existing) {
            await this.prisma.activity.create({
                data: {
                    leadId: existing.id,
                    userId: user.sub,
                    type: shared_1.ActivityType.NOTE,
                    title: 'Duplicate lead attempt spotted',
                    description: `Attempted source: ${data.source}. Did not duplicate.`,
                },
            });
            return existing;
        }
        const lead = await this.prisma.lead.create({
            data: { ...data, phone, tenantId: user.tenantId },
            include: { project: true },
        });
        const assignedToId = await this.autoAssign(user.tenantId);
        if (assignedToId) {
            await this.prisma.lead.update({
                where: { id: lead.id },
                data: { assignedToId },
            });
            await this.notificationsService.create({
                tenantId: user.tenantId,
                userId: assignedToId,
                title: 'New Lead Assigned',
                message: `You have been assigned a new lead: ${lead.name}`,
                type: 'NEW_LEAD',
                leadId: lead.id,
            });
        }
        if (lead.email && !lead.emailOptOut) {
            await this.emailQueue.add('send', {
                to: lead.email,
                subject: `Welcome to ${lead.project?.name || 'Waterting CRM'}`,
                html: `Hi ${lead.name}, thanks for your interest. Our representative will contact you soon.`,
            });
        }
        await this.aiScoringQueue.add('score-lead', { leadId: lead.id, tenantId: user.tenantId });
        return lead;
    }
    async findAll(user, page = 1, limit = 50) {
        return this.prisma.lead.findMany({
            where: { tenantId: user.tenantId },
            take: limit,
            skip: (page - 1) * limit,
            include: {
                project: true,
                assignedTo: { select: { id: true, name: true } },
            },
            orderBy: { createdAt: 'desc' },
        });
    }
    async findOne(user, id) {
        const lead = await this.prisma.lead.findFirst({
            where: { id, tenantId: user.tenantId },
            include: {
                project: true,
                activities: { orderBy: { createdAt: 'desc' } },
                siteVisits: true,
                bookings: true,
            },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        return lead;
    }
    async update(user, id, data) {
        const lead = await this.prisma.lead.findFirst({
            where: { id, tenantId: user.tenantId },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        return this.prisma.lead.update({
            where: { id },
            data,
        });
    }
    async assign(user, id, userId) {
        const lead = await this.prisma.lead.findFirst({
            where: { id, tenantId: user.tenantId },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        const agent = await this.prisma.user.findFirst({
            where: { id: userId, tenantId: user.tenantId },
        });
        if (!agent)
            throw new common_1.NotFoundException('Agent not found');
        return this.prisma.lead.update({
            where: { id },
            data: { assignedToId: userId },
        });
    }
    async addNote(user, id, title, description) {
        const lead = await this.prisma.lead.findFirst({
            where: { id, tenantId: user.tenantId },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        return this.prisma.activity.create({
            data: {
                leadId: id,
                userId: user.sub,
                type: shared_1.ActivityType.NOTE,
                title,
                description,
            },
        });
    }
    async addCall(user, id, duration, outcome) {
        const lead = await this.prisma.lead.findFirst({
            where: { id, tenantId: user.tenantId },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        return this.prisma.activity.create({
            data: {
                leadId: id,
                userId: user.sub,
                type: shared_1.ActivityType.CALL,
                title: `Call logged - ${outcome}`,
                description: `Duration: ${duration}. Outcome: ${outcome}`,
            },
        });
    }
    async updateStage(user, id, stage) {
        const lead = await this.prisma.lead.findUnique({ where: { id } });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        await this.prisma.lead.update({
            where: { id },
            data: { stage: stage },
        });
        await this.prisma.activity.create({
            data: {
                leadId: id,
                userId: user.sub,
                type: shared_1.ActivityType.STAGE_CHANGE,
                title: `Stage changed to ${stage}`,
            },
        });
        this.gateway.emitToTenant(user.tenantId, 'lead:updated', { leadId: id, stage });
        await this.handleStageTrigger(user, lead, stage);
        return lead;
    }
    async handleStageTrigger(user, lead, stage) {
        switch (stage) {
            case 'CONTACTED':
                break;
            case 'INTERESTED':
                if (lead.email && !lead.emailOptOut) {
                    await this.emailQueue.add('send', {
                        to: lead.email,
                        subject: 'Project Brochure',
                        html: `Hi ${lead.name}, please find the brochure attached.`,
                    });
                }
                break;
            case 'VISIT_SCHEDULED':
                if (lead.email && !lead.emailOptOut) {
                    await this.emailQueue.add('send', {
                        to: lead.email,
                        subject: 'Site Visit Scheduled',
                        template: 'visit-confirmation',
                        context: { name: lead.name, date: new Date().toLocaleDateString() },
                    });
                    await this.emailQueue.add('send', {
                        to: lead.email,
                        subject: 'Reminder: Site Visit Tomorrow',
                        html: `Hi ${lead.name}, your visit is scheduled for tomorrow.`,
                    }, { delay: 24 * 3600000 });
                    await this.emailQueue.add('send', {
                        to: lead.email,
                        subject: 'Reminder: Site Visit in 2 Hours',
                        html: `Hi ${lead.name}, your visit starts in 2 hours!`,
                    }, { delay: 22 * 3600000 });
                }
                break;
            case 'VISIT_DONE':
                if (lead.email && !lead.emailOptOut) {
                    await this.emailQueue.add('send', {
                        to: lead.email,
                        subject: 'Thank you for visiting!',
                        html: `Hi ${lead.name}, it was great meeting you today. Let us know if you have questions.`,
                    }, { delay: 4 * 3600000 });
                }
                break;
            case 'NEGOTIATION':
                const admins = await this.prisma.user.findMany({
                    where: { tenantId: user.tenantId, role: { in: ['TENANT_ADMIN', 'SALES_MANAGER'] } },
                });
                for (const admin of admins) {
                    await this.notificationsService.create({
                        tenantId: user.tenantId,
                        userId: admin.id,
                        title: `Negotiation Stage: ${lead.name}`,
                        message: `${lead.assignedTo?.name || 'Agent'} moved ${lead.name} to Negotiation. Close it!`,
                        type: 'NEGOTIATION',
                        leadId: lead.id,
                    });
                }
                break;
            case 'BOOKING_DONE':
                const booking = await this.prisma.booking.findFirst({
                    where: { leadId: lead.id },
                    orderBy: { createdAt: 'desc' },
                    include: { unit: { include: { tower: { include: { project: true } } } } },
                });
                if (booking) {
                    await this.pdfQueue.add('booking-confirmation', { bookingId: booking.id });
                    if (booking.buyerEmail) {
                        await this.emailQueue.add('send', {
                            to: booking.buyerEmail,
                            subject: `Booking Confirmed — ${booking.unit.tower.project.name} Unit ${booking.unit.unitNumber}`,
                            html: `<h2>Congratulations ${booking.buyerName}!</h2><p>Your booking for Unit ${booking.unit.unitNumber} at ${booking.unit.tower.project.name} is confirmed.</p><p><a href="${process.env.FRONTEND_URL}/portal/login">View your booking in the client portal →</a></p>`,
                        });
                    }
                    const accounts = await this.prisma.user.findMany({
                        where: { tenantId: user.tenantId, role: 'ACCOUNTS' },
                    });
                    for (const acc of accounts) {
                        await this.notificationsService.create({
                            tenantId: user.tenantId,
                            userId: acc.id,
                            title: `New Booking: ${booking.buyerName}`,
                            message: `Unit ${booking.unit.unitNumber} — ₹${booking.bookingAmount.toLocaleString('en-IN')} — document collection needed in 48h`,
                            type: 'BOOKING',
                            leadId: lead.id,
                        });
                    }
                }
                break;
            case 'LOST':
                if (lead.email && !lead.emailOptOut) {
                    const delays = [90, 120, 180].map(d => d * 24 * 3600000);
                    for (const d of delays) {
                        await this.emailQueue.add('send', {
                            to: lead.email,
                            subject: 'Checking in',
                            html: `Hi ${lead.name}, we haven't spoken in a while. Any updates on your home search?`,
                        }, { delay: d });
                    }
                }
                break;
        }
    }
    async autoAssign(tenantId) {
        const agents = await this.prisma.user.findMany({
            where: { tenantId, role: 'SALES_AGENT', isActive: true },
            include: { _count: { select: { leads: true } } },
            orderBy: { leads: { _count: 'asc' } },
        });
        return agents[0]?.id ?? null;
    }
    async createFromWebhook(data) {
        const tenantId = data.tenantId || (await this.prisma.tenant.findFirst())?.id;
        if (!tenantId)
            throw new Error('No tenant context found for webhook');
        const phone = this.normalizePhone(data.phone);
        const existing = await this.prisma.lead.findFirst({
            where: { tenantId, phone },
        });
        if (existing)
            return existing;
        const lead = await this.prisma.lead.create({
            data: {
                name: data.name,
                phone,
                email: data.email,
                source: data.source || shared_1.LeadSource.WEBSITE,
                tenantId,
            },
        });
        await this.aiScoringQueue.add('score-lead', { leadId: lead.id, tenantId });
        return lead;
    }
};
exports.LeadsService = LeadsService;
exports.LeadsService = LeadsService = __decorate([
    (0, common_1.Injectable)(),
    __param(3, (0, bull_1.InjectQueue)('ai-scoring')),
    __param(4, (0, bull_1.InjectQueue)('email')),
    __param(5, (0, bull_1.InjectQueue)('pdf')),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        events_gateway_1.EventsGateway,
        notifications_service_1.NotificationsService, Object, Object, Object])
], LeadsService);
