// api/dist/apps/api/src/modules/site-visits/site-visits.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SiteVisitsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
const shared_1 = require("@waterting/shared");
let SiteVisitsService = class SiteVisitsService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async schedule(user, data) {
        const lead = await this.prisma.lead.findFirst({
            where: { id: data.leadId, tenantId: user.tenantId },
        });
        if (!lead)
            throw new common_1.NotFoundException('Lead not found');
        const visit = await this.prisma.siteVisit.create({
            data: {
                ...data,
            },
        });
        await this.prisma.activity.create({
            data: {
                leadId: data.leadId,
                userId: user.sub,
                type: shared_1.ActivityType.VISIT_SCHEDULED,
                title: 'Site Visit Scheduled',
                metadata: { scheduledAt: data.scheduledAt },
            },
        });
        return visit;
    }
    async recordOutcome(user, id, outcome, notes) {
        const visit = await this.prisma.siteVisit.findUnique({
            where: { id },
            include: { lead: true },
        });
        if (!visit || visit.lead.tenantId !== user.tenantId) {
            throw new common_1.NotFoundException('Site Visit not found');
        }
        const updated = await this.prisma.siteVisit.update({
            where: { id },
            data: { outcome, notes },
        });
        await this.prisma.activity.create({
            data: {
                leadId: visit.lead.id,
                userId: user.sub,
                type: shared_1.ActivityType.VISIT_COMPLETED,
                title: `Site Visit Completed: ${outcome}`,
                description: notes,
            },
        });
        return updated;
    }
    async findAll(user) {
        return this.prisma.siteVisit.findMany({
            where: { lead: { tenantId: user.tenantId } },
            include: { lead: true, agent: true },
            orderBy: { scheduledAt: 'asc' },
        });
    }
};
exports.SiteVisitsService = SiteVisitsService;
exports.SiteVisitsService = SiteVisitsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], SiteVisitsService);
