// api/dist/apps/api/src/app.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppModule = void 0;
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const app_controller_1 = require("./app.controller");
const app_service_1 = require("./app.service");
const prisma_module_1 = require("./common/prisma/prisma.module");
const auth_module_1 = require("./modules/auth/auth.module");
const jwt_auth_guard_1 = require("./common/guards/jwt-auth.guard");
const roles_guard_1 = require("./common/guards/roles.guard");
const tenants_module_1 = require("./modules/tenants/tenants.module");
const users_module_1 = require("./modules/users/users.module");
const projects_module_1 = require("./modules/projects/projects.module");
const leads_module_1 = require("./modules/leads/leads.module");
const towers_module_1 = require("./modules/towers/towers.module");
const units_module_1 = require("./modules/units/units.module");
const webhooks_module_1 = require("./modules/webhooks/webhooks.module");
const site_visits_module_1 = require("./modules/site-visits/site-visits.module");
const brokers_module_1 = require("./modules/brokers/brokers.module");
const bookings_module_1 = require("./modules/bookings/bookings.module");
const payments_module_1 = require("./modules/payments/payments.module");
const media_module_1 = require("./modules/media/media.module");
const dashboard_module_1 = require("./modules/dashboard/dashboard.module");
const analytics_module_1 = require("./modules/analytics/analytics.module");
const automations_module_1 = require("./modules/automations/automations.module");
const portal_module_1 = require("./modules/portal/portal.module");
const activities_module_1 = require("./modules/activities/activities.module");
const notifications_module_1 = require("./modules/notifications/notifications.module");
const bull_1 = require("@nestjs/bull");
const schedule_1 = require("@nestjs/schedule");
const workers_module_1 = require("./workers/workers.module");
let AppModule = class AppModule {
};
exports.AppModule = AppModule;
exports.AppModule = AppModule = __decorate([
    (0, common_1.Module)({
        imports: [
            schedule_1.ScheduleModule.forRoot(),
            bull_1.BullModule.forRoot({
                redis: process.env.REDIS_URL,
            }),
            bull_1.BullModule.registerQueue({ name: 'ai-scoring' }, { name: 'email' }, { name: 'pdf' }, { name: 'portal-sync' }),
            prisma_module_1.PrismaModule,
            auth_module_1.AuthModule,
            tenants_module_1.TenantsModule,
            users_module_1.UsersModule,
            projects_module_1.ProjectsModule,
            leads_module_1.LeadsModule,
            towers_module_1.TowersModule,
            units_module_1.UnitsModule,
            webhooks_module_1.WebhooksModule,
            site_visits_module_1.SiteVisitsModule,
            brokers_module_1.BrokersModule,
            bookings_module_1.BookingsModule,
            payments_module_1.PaymentsModule,
            media_module_1.MediaModule,
            dashboard_module_1.DashboardModule,
            analytics_module_1.AnalyticsModule,
            automations_module_1.AutomationsModule,
            portal_module_1.PortalModule,
            activities_module_1.ActivitiesModule,
            notifications_module_1.NotificationsModule,
            workers_module_1.WorkersModule,
        ],
        controllers: [app_controller_1.AppController],
        providers: [
            app_service_1.AppService,
            {
                provide: core_1.APP_GUARD,
                useClass: jwt_auth_guard_1.JwtAuthGuard,
            },
            {
                provide: core_1.APP_GUARD,
                useClass: roles_guard_1.RolesGuard,
            },
        ],
    })
], AppModule);
