// api/dist/apps/api/src/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@nestjs/core");
const app_module_1 = require("./app.module");
const common_1 = require("@nestjs/common");
async function bootstrap() {
    const app = await core_1.NestFactory.create(app_module_1.AppModule);
    const logger = new common_1.Logger('Bootstrap');
    app.enableCors({
        origin: (origin, callback) => {
            const allowed = [
                undefined,
                'http://localhost:3000',
                process.env.FRONTEND_URL,
                /waterting.*\.vercel\.app$/,
            ];
            const ok = allowed.some(a => !a ? !origin : a instanceof RegExp ? a.test(origin ?? '') : a === origin);
            ok ? callback(null, true) : callback(new Error(`CORS blocked: ${origin}`));
        },
        methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'],
        allowedHeaders: ['Content-Type', 'Authorization', 'Accept'],
        credentials: true,
    });
    app.useGlobalPipes(new common_1.ValidationPipe({
        whitelist: true,
        transform: true,
        transformOptions: { enableImplicitConversion: true },
    }));
    app.use((req, res, next) => {
        res.on('finish', () => {
            const level = res.statusCode >= 400 ? 'warn' : 'log';
            logger[level](`${req.method} ${req.url} → ${res.statusCode}`);
        });
        next();
    });
    await app.listen(process.env.PORT ?? 3001);
    logger.log(`API running on port ${process.env.PORT ?? 3001}`);
}
bootstrap();
