// api/dist/apps/api/src/modules/projects/construction.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructionService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
let ConstructionService = class ConstructionService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async createUpdate(user, projectId, data) {
        return this.prisma.constructionUpdate.create({
            data: {
                ...data,
                projectId,
            },
        });
    }
    async getProjectFeed(projectId) {
        return this.prisma.constructionUpdate.findMany({
            where: { projectId },
            orderBy: { updateDate: 'desc' },
        });
    }
    async getBuyerFeed(buyerEmail) {
        const booking = await this.prisma.booking.findFirst({
            where: { buyerEmail },
            include: { unit: { include: { tower: true } } },
        });
        if (!booking)
            throw new common_1.NotFoundException('No booking found');
        return this.getProjectFeed(booking.unit.tower.projectId);
    }
};
exports.ConstructionService = ConstructionService;
exports.ConstructionService = ConstructionService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], ConstructionService);
