// api/dist/apps/api/src/modules/projects/projects.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
let ProjectsService = class ProjectsService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async findAll(user, page = 1, limit = 50) {
        return this.prisma.project.findMany({
            where: { tenantId: user.tenantId },
            take: limit,
            skip: (page - 1) * limit,
            include: { towers: true },
            orderBy: { createdAt: 'desc' },
        });
    }
    async findOne(user, id) {
        const project = await this.prisma.project.findFirst({
            where: { id, tenantId: user.tenantId },
            include: { towers: { include: { units: true } } },
        });
        if (!project)
            throw new common_1.NotFoundException('Project not found');
        return project;
    }
    async create(user, data) {
        return this.prisma.project.create({
            data: {
                ...data,
                tenantId: user.tenantId,
            },
        });
    }
    async update(user, id, data) {
        return this.prisma.project.updateMany({
            where: { id, tenantId: user.tenantId },
            data,
        });
    }
    async remove(user, id) {
        return this.prisma.project.deleteMany({
            where: { id, tenantId: user.tenantId },
        });
    }
};
exports.ProjectsService = ProjectsService;
exports.ProjectsService = ProjectsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], ProjectsService);
