// api/dist/apps/api/src/modules/analytics/analytics.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticsService = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../common/prisma/prisma.service");
let AnalyticsService = class AnalyticsService {
    constructor(prisma) {
        this.prisma = prisma;
    }
    async getInventoryInsights(tenantId, projectId) {
        const units = await this.prisma.unit.findMany({
            where: projectId
                ? { tower: { projectId } }
                : { tower: { project: { tenantId } } },
            select: { status: true, totalPrice: true },
        });
        const counts = {
            TOTAL: units.length,
            AVAILABLE: units.filter(u => u.status === 'AVAILABLE').length,
            BOOKED: units.filter(u => u.status === 'BOOKED').length,
            SOLD: units.filter(u => u.status === 'SOLD').length,
            RESERVED: units.filter(u => u.status === 'RESERVED').length,
        };
        const salesValue = units
            .filter(u => u.status === 'BOOKED' || u.status === 'SOLD')
            .reduce((acc, u) => acc + u.totalPrice, 0);
        const bookings = await this.prisma.booking.count({
            where: { lead: { tenantId }, createdAt: { gte: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000) } },
        });
        return {
            statusBreakdown: counts,
            totalSalesValue: salesValue,
            absorptionRate30d: bookings,
            occupancyPct: Math.round(((counts.TOTAL - counts.AVAILABLE) / counts.TOTAL) * 100) || 0,
        };
    }
};
exports.AnalyticsService = AnalyticsService;
exports.AnalyticsService = AnalyticsService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService])
], AnalyticsService);
